src/main/java/edu/byu/cs/util/RepoUrlValidator.java

```diff
@@ -0,0 +1,106 @@
+package edu.byu.cs.util;
+
+import edu.byu.cs.autograder.GradingException;
+import org.eclipse.jgit.annotations.Nullable;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RepoUrlValidator {
+
+    public static boolean isValid(@Nullable String repoUrl) {
+        try {
+            clean(repoUrl);
+            return true;
+        } catch (InvalidRepoUrlException e) {
+            return false;
+        }
+    }
+
+    public static String clean(@Nullable String repoUrl) throws InvalidRepoUrlException {
+        return new RepoUrlValidator().cleanRepoUrl(repoUrl);
+    }
+
+    /**
+     * Cleans and standardizes the student's repo URL.
+     * <br>
+     * The resulting string can be used to <code>git clone</code> the repo.
+     * The standardized URL is also a suitable identifier for repository uniqueness comparisons.
+     *
+     * @param repoUrl The student's repository URL.
+     * @return Cleaned and standardized repository URL.
+     * @throws InvalidRepoUrlException If the repo URL is invalid.
+     */
+    public String cleanRepoUrl(@Nullable String repoUrl) throws InvalidRepoUrlException {
+        if (repoUrl == null) {
+            throw new InvalidRepoUrlException("NULL is not a valid repo URL.");
+        }
+        String trimmedRepoUrl = repoUrl.trim();
+
+        Pattern pattern;
+        Matcher matcher;
+
+        String domainName;
+        String githubUsername;
+        String repositoryName;
+        for (String regexPattern : getRepoRegexPatterns()) {
+            pattern = Pattern.compile(regexPattern, Pattern.CASE_INSENSITIVE);
+            matcher = pattern.matcher(trimmedRepoUrl);
+            if (matcher.find()) {
+                domainName = matcher.group(1).toLowerCase();
+                githubUsername = matcher.group(2);
+                repositoryName = matcher.group(3);
+                return String.format("https://%s/%s/%s", domainName, githubUsername, repositoryName);
```

Comment:
A more generic solution would abstract this exact url formatter, and ideally it would associate the extracted with the formatter. That would keep those too closely related pieces of functionality coupled together as they should be.

Reply:
This fits our current use case just fine. If we need to change this, we'll need to modify the source code anyways. This is not following the open-closed principle, but we can accept that for now.